# Melbourne Housing Market

Dataset referente ao mercado imobiliario de Melbourne.(preciso melhorar a descrição)


Dataset encontra-se disponivel no endereço: https://www.kaggle.com/datasets/anthonypino/melbourne-housing-market

In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.linear_model import LinearRegression
ols = LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.metrics import mean_absolute_error 
from sklearn.metrics import mean_squared_error 
from sklearn.metrics import r2_score 

In [3]:
df = pd.read_csv(r"C:\Users\dougl\OneDrive\Área de Trabalho\Projetos GitHub\Melbourne_house\archive\Melbourne_housing_FULL.csv")

In [4]:
df.head(5)

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [5]:
df.sample(5)

Suburb           Address  Rooms Type      Price Method   SellerG  \
3860         Elwood   3/307 Barkly St      2    u        NaN     PI    Biggin   
18250  Princes Hill      78 Arnold St      3    h  1540000.0      S   Collins   
19751      Armadale    121 Kooyong Rd      4    h  5100000.0      S  Marshall   
28101     Reservoir  1/11 Mahoneys Rd      3    u        NaN     PI       Ray   
33443       Ivanhoe     2/16 Green St      2    t   755000.0      S    Nelson   

             Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
3860   10/12/2016       7.7    3184.0  ...       1.0  1.0     565.0   
18250  12/08/2017       3.5    3054.0  ...       2.0  1.0     288.0   
19751   3/09/2017       6.3    3143.0  ...       2.0  2.0     761.0   
28101  18/11/2017      12.0    3073.0  ...       NaN  NaN       NaN   
33443  17/03/2018       7.8    3079.0  ...       NaN  NaN       NaN   

       BuildingArea  YearBuilt                CouncilArea Lattitude  \
3860           86.0     1930.0  Port Phillip City Council -37.87460   
18250           NaN        NaN         Yarra City Council -37.78311   
19751         255.0     1900.0   Stonnington City Council -37.85249   
28101           NaN        NaN       Darebin City Council       NaN   
33443           NaN        NaN       Banyule City Council       NaN   

       Longtitude             Regionname Propertycount  
3860    144.97960  Southern Metropolitan        8989.0  
18250   144.96548  Northern Metropolitan        1008.0  
19751   145.02091  Southern Metropolitan        4836.0  
28101         NaN  Northern Metropolitan       21650.0  
33443         NaN   Eastern Metropolitan        5549.0  

[5 rows x 21 columns]

In [6]:
df.tail(5)

Suburb           Address  Rooms Type      Price Method  \
34852  Yarraville       13 Burns St      4    h  1480000.0     PI   
34853  Yarraville     29A Murray St      2    h   888000.0     SP   
34854  Yarraville    147A Severn St      2    t   705000.0      S   
34855  Yarraville  12/37 Stephen St      3    h  1140000.0     SP   
34856  Yarraville  3 Tarrengower St      2    h  1020000.0     PI   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
34852            Jas  24/02/2018       6.3    3013.0  ...       1.0  3.0   
34853        Sweeney  24/02/2018       6.3    3013.0  ...       2.0  1.0   
34854            Jas  24/02/2018       6.3    3013.0  ...       1.0  2.0   
34855  hockingstuart  24/02/2018       6.3    3013.0  ...       NaN  NaN   
34856             RW  24/02/2018       6.3    3013.0  ...       1.0  0.0   

       Landsize  BuildingArea  YearBuilt               CouncilArea Lattitude  \
34852     593.0           NaN        NaN  Maribyrnong City Council -37.81053   
34853      98.0         104.0     2018.0  Maribyrnong City Council -37.81551   
34854     220.0         120.0     2000.0  Maribyrnong City Council -37.82286   
34855       NaN           NaN        NaN  Maribyrnong City Council       NaN   
34856     250.0         103.0     1930.0  Maribyrnong City Council -37.81810   

       Longtitude            Regionname Propertycount  
34852   144.88467  Western Metropolitan        6543.0  
34853   144.88826  Western Metropolitan        6543.0  
34854   144.87856  Western Metropolitan        6543.0  
34855         NaN  Western Metropolitan        6543.0  
34856   144.89351  Western Metropolitan        6543.0  

[5 rows x 21 columns]

In [7]:
df.shape

(34857, 21)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

In [9]:
df.duplicated().sum()

1

In [10]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

É possivel visualizar acima, os dados que apresentão algum tipo de campo nulo. será iniado o tratamento apartir destes dados.

Para estimar um valor, é necessario que alguns campos tenha obrigatoriamente valores. Caso não tenha, esse dado sera descartado.
* Bedroom2 possui 8217 campos nulos.
* Bathroom possui 8226 campos nulos.
* Price possui 7610 campos nulos.


In [11]:
#df.dropna(subset=["Bedroom2"],inplace=True)
#df.dropna(subset=["Bathroom"],inplace=True)
#df.dropna(subset=["Price"],inplace=True)
#df.dropna(subset=["BuildingArea"],inplace=True)


In [12]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

Precisa verificar endereços duplicados

In [156]:
df["Address"].duplicated().sum()

848

In [141]:
df_teste_area = df.loc[(df['Bathroom']== 1)&(df['Bedroom2']== 3)&(df['CouncilArea']=='Boroondara City Council')&(df['Suburb']=="Glen Iris")&(df['Address']=='48 Alfred Rd')]

In [142]:
df_teste_area['BuildingArea'].isnull().sum()

0

In [143]:
df_teste_area["Suburb"].value_counts()

Glen Iris    2
Name: Suburb, dtype: int64

In [144]:
df_teste_area["Address"].value_counts()

48 Alfred Rd    2
Name: Address, dtype: int64

In [132]:
df["Bathroom"].value_counts()

1.0     12969
2.0     11064
3.0      2181
4.0       269
5.0        77
0.0        46
6.0        16
7.0         4
8.0         3
12.0        1
9.0         1
Name: Bathroom, dtype: int64

In [111]:
df["Bedroom2"].value_counts()

3.0     11881
4.0      6348
2.0      5777
5.0      1427
1.0       966
6.0       168
7.0        30
0.0        17
8.0        13
9.0         5
10.0        4
30.0        1
20.0        1
16.0        1
12.0        1
Name: Bedroom2, dtype: int64

In [127]:
df["Suburb"].value_counts()

Reservoir          844
Bentleigh East     583
Richmond           552
Glen Iris          491
Preston            485
                  ... 
Belgrave             1
Cranbourne East      1
Yarra Glen           1
Olinda               1
Menzies Creek        1
Name: Suburb, Length: 351, dtype: int64

In [112]:
df["Regionname"].value_counts()

Southern Metropolitan         11836
Northern Metropolitan          9557
Western Metropolitan           6799
Eastern Metropolitan           4377
South-Eastern Metropolitan     1739
Eastern Victoria                228
Northern Victoria               203
Western Victoria                115
Name: Regionname, dtype: int64

In [113]:
df["CouncilArea"].value_counts()

Boroondara City Council           3675
Darebin City Council              2851
Moreland City Council             2122
Glen Eira City Council            2006
Melbourne City Council            1952
Banyule City Council              1861
Moonee Valley City Council        1791
Bayside City Council              1764
Brimbank City Council             1593
Monash City Council               1466
Stonnington City Council          1460
Maribyrnong City Council          1451
Port Phillip City Council         1280
Hume City Council                 1214
Yarra City Council                1186
Manningham City Council           1046
Hobsons Bay City Council           942
Kingston City Council              871
Whittlesea City Council            828
Wyndham City Council               624
Whitehorse City Council            618
Maroondah City Council             506
Knox City Council                  371
Greater Dandenong City Council     314
Melton City Council                292
Frankston City Council   

In [145]:
df_teste_area

Suburb       Address  Rooms Type      Price Method SellerG  \
4522   Glen Iris  48 Alfred Rd      3    h  1530000.0      S  Jellis   
25970  Glen Iris  48 Alfred Rd      3    h        NaN     SP  Jellis   

             Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
4522    4/06/2016       9.2    3146.0  ...       1.0  1.0     753.0   
25970  28/10/2017       7.3    3146.0  ...       1.0  2.0     754.0   

       BuildingArea  YearBuilt              CouncilArea Lattitude  Longtitude  \
4522          118.0     1940.0  Boroondara City Council -37.85430   145.08890   
25970         118.0     1940.0  Boroondara City Council -37.85431   145.08889   

                  Regionname Propertycount  
4522   Southern Metropolitan       10412.0  
25970  Southern Metropolitan       10412.0  

[2 rows x 21 columns]

In [115]:
df[df["CouncilArea"].isnull()]

Suburb           Address  Rooms Type     Price Method SellerG  \
18523    Footscray  2/16 Stafford St      2    u  710000.0      S     Jas   
26888   Camberwell   2/3 Kingsley St      2    h  825000.0     VB  Jellis   
29483  Fawkner Lot      1/3 Brian St      3    h  616000.0     SP    Brad   

             Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
18523  15/07/2017       5.1    3011.0  ...       NaN  NaN       NaN   
26888  11/11/2017       7.7    3124.0  ...       NaN  NaN       NaN   
29483   6/01/2018       NaN       NaN  ...       NaN  NaN       NaN   

       BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  Regionname  \
18523           NaN        NaN          NaN       NaN         NaN         NaN   
26888           NaN        NaN          NaN       NaN         NaN         NaN   
29483           NaN        NaN          NaN       NaN         NaN         NaN   

      Propertycount  
18523           NaN  
26888           NaN  
29483           NaN  

[3 rows x 21 columns]

Nao fazer alteraçoes abaixo

In [116]:
X = df.drop(["Price","Car", "Suburb","Distance","Postcode","Longtitude", "Propertycount","Lattitude", "Address", "Type", "Method", "YearBuilt", "SellerG", "BuildingArea", "Date", "CouncilArea", "Regionname"], axis=1)

In [117]:
Y = df["Price"]

In [118]:
X

Rooms  Bedroom2  Bathroom  Landsize
0          2       2.0       1.0     126.0
1          2       2.0       1.0     202.0
2          2       2.0       1.0     156.0
3          3       3.0       2.0       0.0
4          3       3.0       2.0     134.0
...      ...       ...       ...       ...
34852      4       4.0       1.0     593.0
34853      2       2.0       2.0      98.0
34854      2       2.0       1.0     220.0
34855      3       NaN       NaN       NaN
34856      2       2.0       1.0     250.0

[34857 rows x 4 columns]

In [119]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Rooms     34857 non-null  int64  
 1   Bedroom2  26640 non-null  float64
 2   Bathroom  26631 non-null  float64
 3   Landsize  23047 non-null  float64
dtypes: float64(3), int64(1)
memory usage: 1.1 MB


In [120]:
Y

0              NaN
1        1480000.0
2        1035000.0
3              NaN
4        1465000.0
           ...    
34852    1480000.0
34853     888000.0
34854     705000.0
34855    1140000.0
34856    1020000.0
Name: Price, Length: 34857, dtype: float64

In [121]:
Y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 34857 entries, 0 to 34856
Series name: Price
Non-Null Count  Dtype  
--------------  -----  
27247 non-null  float64
dtypes: float64(1)
memory usage: 272.4 KB


In [122]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.28, random_state=42)

NameError: name 'y' is not defined

In [ ]:
X.shape, y.shape, X_train.shape, X_test.shape, y_train.shape, y_test.shape

((13580, 4), (13580,), (9777, 4), (3803, 4), (9777,), (3803,))

In [ ]:
ols.fit(X,Y)

LinearRegression()

In [ ]:
ols.intercept_

31584.605965678697

In [ ]:
ols.coef_

array([2.24216833e+05, 2.53683552e+03, 2.45242799e+05, 3.03574181e+00])

In [ ]:
predito = ols.predict(X)
predito

array([ 730947.96150257,  730808.31737929, 1202737.99820381, ...,
       1203654.79223056, 1186471.03115835, 1184941.0172859 ])

In [ ]:
mean_absolute_error(y, predito) 

377922.38368053216

In [ ]:
mean_squared_error(y, predito)

288970799009.69965

In [ ]:
r2_score(y, predito)

0.29293073279582105